lib/active_record_enum.rb

```diff
@@ -0,0 +1,235 @@
+
+          pairs = values.respond_to?(:each_pair) ? values.each_pair : values.each_with_index
+          pairs.each do |value, i|
+            enum_values[value] = i
+
+            # def active?() status == 0 end
+            klass.send(:detect_enum_conflict!, name, "#{value}?")
+            define_method("#{value}?") { self[name] == i }
+
+            # def active!() update! status: :active end
+            klass.send(:detect_enum_conflict!, name, "#{value}!")
+            define_method("#{value}!") { update! name => value }
+
+            # scope :active, -> { where status: 0 }
+            klass.send(:detect_enum_conflict!, name, value, true)
+            klass.scope value, -> { klass.where name => i }
+          end
+        end
+        defined_enums[name.to_s] = enum_values
+      end
+    end
+
+    private
+      def _enum_methods_module
+        @_enum_methods_module ||= begin
+          mod = Module.new do
+            private
+              def save_changed_attribute(attr_name, old)
+                if (mapping = self.class.defined_enums[attr_name.to_s])
+                  value = _read_attribute(attr_name)
+                  if attribute_changed?(attr_name)
+                    if mapping[old] == value
+                      clear_attribute_changes([attr_name])
+                    end
+                  else
+                    if old != value
+                      set_attribute_was(attr_name, mapping.key(old))
+                    end
+                  end
+                else
+                  super
+                end
+              end
+          end
+          include mod
+          mod
+        end
+      end
+
+      ENUM_CONFLICT_MESSAGE = \
+        "You tried to define an enum named \"%{enum}\" on the model \"%{klass}\", but " \
+        "this will generate a %{type} method \"%{method}\", which is already defined " \
+        "by %{source}."
+
+      def detect_enum_conflict!(enum_name, method_name, klass_method = false)
+        if klass_method && dangerous_class_method?(method_name)
+          raise ArgumentError, ENUM_CONFLICT_MESSAGE % {
+            enum: enum_name,
+            klass: self.name,
+            type: 'class',
+            method: method_name,
+            source: 'Active Record'
+          }
+        elsif !klass_method && dangerous_attribute_method?(method_name)
+          raise ArgumentError, ENUM_CONFLICT_MESSAGE % {
+            enum: enum_name,
+            klass: self.name,
+            type: 'instance',
+            method: method_name,
+            source: 'Active Record'
+          }
+        elsif !klass_method && method_defined_within?(method_name, _enum_methods_module, Module)
+          raise ArgumentError, ENUM_CONFLICT_MESSAGE % {
+            enum: enum_name,
+            klass: self.name,
+            type: 'instance',
+            method: method_name,
+            source: 'another enum'
+          }
+        end
```

Comment:
This is a neat use of `%` to fill in the `ENUM_CONFLICT_MESSAGE` messages.  It's really just `sprintf` but I like how it reads. I know, you just copied it in :)